src/Cluster.jl

```diff
@@ -27,31 +33,91 @@ function pcca(transition_matrix::Matrix{Float64})
+    return assignments, num_clusters
 end
 
-# TODO: replace placeholder simple cluster membership function with more sophisticated function
-function assign_cluster_membership(n_states, n_clusters, eigenvectors)
+# compute membership distributions probabilities that the microstates belong to the same metastable state
+# by using the properties of slow processes in eigenvector space
+function assign_cluster_membership(n_states, num_clusters, eigenvectors)
     assignments = zeros(Int, n_states)
     for i in 1:n_states
         # assign state i to cluster with largest component in eigenvector matrix
-        _, index = findmax(abs.(eigenvectors[i, 2:n_clusters+1]))
+        _, index = findmax(abs.(eigenvectors[i, 2:num_clusters+1]))
         assignments[i] = index
     end
     return assignments
 end
 
-# If computeEigen performance isn't sufficient, use optimized eigen decomposition on sparse matrix with Arpack eigs
-# num_eigenvalues: and only compute a certain number of eigenvalues & eigenvectors
-# Cmputes eigenvalues of largest magnitude by default. See documentation: https://arpack.julialinearalgebra.org/stable/eigs/
-function compute_sparse_eigen(matrix::SparseMatrixCSC, num_eigenvalues::Int=3)
-    # converts to a sparse matrix (SparseMatrixCSC)
-    sparse_matrix = sparse(matrix)
-    eigen = eigs(sparse_matrix, nev=num_eigenvalues)
-    eigenvalues = eigen[1]
-    eigenvectors = eigen[2]
-    return eigenvalues, eigenvectors
-end
\ No newline at end of file
+# Reimplementation of DeepTime PCCA in Julia
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/_pcca.py#L9
+function pcca(P, num_clusters)
+    pi = compute_stationary_distribution(P)
+    memberships = pcca_memberships(P, num_clusters, pi)
+    pi_coarse_grained = memberships' * pi  # coarse-grained stationary distribution
+
+    # HMM output matrix
+    HMM_output_mat = Diagonal(1.0 ./ pi_coarse_grained) * memberships' * Diagonal(pi)
+
+    # renormalize B to make it row-stochastic
+    HMM_output_mat .= HMM_output_mat ./ sum(HMM_output_mat, dims=2)
+
+    # coarse-grained transition matrix
+    W = inv(memberships' * memberships)
+    A = memberships' * P * memberships
+    P_coarse_grained = W * A
+
+    # symmetrize & renormalize (eliminates numerical errors)
+    X = Diagonal(pi_coarse_grained) * P_coarse_grained
+    
+    # normalize to get the coarse-grained transition matrix
+    P_coarse_grained = X ./ sum(X, dims=2)
+
+    return PCCAModel(P_coarse_grained, pi_coarse_grained, memberships, HMM_output_mat)
+end
+
+# Reimplementation of DeepTime PCCAModel in Julia
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/_pcca.py#L71
+abstract type Model end
+
+struct PCCAModel <: Model
+    transition_matrix_coarse::Matrix{Float64}
+    pi_coarse::Vector{Float64}
+    memberships::Matrix{Float64}
+    metastable_distributions::Matrix{Float64}
+    m::Int
+end
```

Comment:
It might be better to parametrize this Struct by using a type parameter `T` instead of the hard-coded `Float64`

Reply:
fixed in 3147d8c3ac1b28d5b886e16dcf4a05d3b2497d16